File: 2015/Day5/src/day5.py
import sys
debug = False

def check_nice(line : str)->int:
    if(debug):
        print(f"Line: {line}")
    num_vowels = 0
    for vowel in ['a','e','i','o','u']:
        if vowel in line:
            num_vowels += line.count(vowel)
    last_char = ''
    has_consecutive_chars = False
    for char in line:
        if char == last_char:
            has_consecutive_chars = True
            break
        last_char = char
    has_bad_strings = False
    bad_strings = ['ab', 'cd', 'pq', 'xy']
    for bad_string in bad_strings:
        if bad_string in line:
            has_bad_strings = True
            break
    if(debug):
        print(f"{has_bad_strings=}, {has_consecutive_chars=}, {num_vowels=}")
    if(not has_bad_strings and has_consecutive_chars and num_vowels >= 3):
        return 1
    return 0

def check_other_nice(line : str) -> int:
    if(debug):
        print(f"{line=}")
    has_multiple_nonconsecutive = False
    for i in range(len(line) - 3):
        first_pair = line[i] + line[i+1]
        for j in range(i + 2, len(line) - 1):
            second_pair = line[j] + line[j+1]
            if(first_pair == second_pair):
                has_multiple_nonconsecutive = True
                if(debug):
                    print(f"{first_pair=} {i=}, {second_pair=} {j=}")
                break
        if(has_multiple_nonconsecutive):
            break

    has_sandwich_letters = False
    for i in range(len(line) - 2):
        if line[i] == line[i + 2]:
            has_sandwich_letters = True
            break

    if(debug):
        print(f"{has_sandwich_letters=}, {has_multiple_nonconsecutive=}\n")
    if(has_sandwich_letters and has_multiple_nonconsecutive):
        return 1
    return 0

if __name__ == "__main__":
    if(len(sys.argv) != 2):
        print("Usage: python3 day5.py input.txt")
        exit(1)
    num_nice_part1 = 0
    num_nice_part2 = 0
    with open(sys.argv[1], "r") as fileobj:
        for line in fileobj:
            num_nice_part1 += check_nice(line.rstrip())
            num_nice_part2 += check_other_nice(line.rstrip())
    print(f"Answer to part 1 is {num_nice_part1}")
    print(f"Answer to part 2 is {num_nice_part2}")

File: 2015/Day5/src/day5.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

#define DEBUG false
#define NUM_VOWELS 5

using namespace std;

int count(string haystack, char needle)
{
    int num_occurences = 0;
    int index = haystack.find(needle);
    while(index != haystack.npos && index < haystack.size())
    {
        num_occurences++;
        index = haystack.find(needle, index + 1);
    }
    return num_occurences;
}

int part1(vector <string> lines)
{
    int num_nice = 0;
    char vowels[NUM_VOWELS] = {'a', 'e', 'i', 'o', 'u'}; 
    for(string line : lines)
    {
        //condition 1
        int vowel_count = 0;
        for(int i = 0; i < NUM_VOWELS; i++)
        {
            vowel_count += count(line, vowels[i]);
        }
        //condition 2
        bool has_consecutive_chars = false;
        char last_char = '\0';
        for(int i = 0; i < line.size(); i++)
        {
            if(line.at(i) == last_char)
            {
                has_consecutive_chars = true;
                break;
            }
            last_char = line.at(i);
        }
        //condition 3
        bool has_bad_strings = false;
        string bad_strings[4] = {"ab", "cd", "pq", "xy"};
        for(string bad_string : bad_strings)
        {
            if(line.find(bad_string) != line.npos)
            {
                has_bad_strings = true;
                break;
            }
        }
        if(DEBUG)
        {
            cout << line << " : " << has_bad_strings << has_consecutive_chars << vowel_count << endl;
        }
        if(!has_bad_strings && has_consecutive_chars && vowel_count >= 3)
        {
            num_nice++;
        }
    }
    return num_nice;
}

int part2(vector <string> lines)
{
    int num_nice = 0;
    for(string line : lines)
    {
        //condition 1
        bool has_multiple_nonconsecutive = false;
        for(int i = 0; i < line.size() - 3; i++)
        {
            string first_pair = line.substr(i, 2);
            for(int j = i + 2; j < line.size() - 1; j++)
            {
                string second_pair = line.substr(j, 2);
                if(first_pair == second_pair)
                {
                    has_multiple_nonconsecutive = true;
                    break;
                }
            } 
            if(has_multiple_nonconsecutive)
            {
                break;
            }
        }
        //condition 2
        bool has_sandwich_letters = false;
        for(int i = 0; i < line.size() - 2; i++)
        {
            if(line.at(i) == line.at(i + 2))
            {
                has_sandwich_letters = true;
                break;
            }
        }
        if(DEBUG)
        {
            cout << line << " : " << has_sandwich_letters << has_multiple_nonconsecutive << endl;
        }
        if(has_sandwich_letters && has_multiple_nonconsecutive)
        {
            num_nice++;
        }
    }
    return num_nice;
}

int main(int argc, char * argv[])
{
    if(argc != 2)
    {
        cout << "Usage: ./day5 input.txt\n";
        exit(1);
    }
    ifstream file(argv[1]);
    vector <string> lines;
    string line;
    getline(file, line);
    while(line.size() > 0)
    {
        lines.push_back(line);
        getline(file, line);
    } 
    file.close();

    //part 1
    int ret_val = part1(lines);
    cout << "The answer to part 1 is " << ret_val << endl;

    //part 2
    ret_val = part2(lines);
    cout << "The answer to part 2 is " << ret_val << endl;
}
